### ISSUE TITLE
[InstallAppleCertificate] Unable to find the certificate SHA1 hash and common name (CN). Verify that the certificate is a valid P12. (Regression, MacOS 13.0)

### ISSUE THREAD

**Issue by fntc** [microsoft/azure-pipelines-tasks#17084]
After installing MacOS 13.0 Ventura on the build machine the InstallAppleCertificate Task fails.
Error is the same as in an old bug.

It seems that Apple now distributes a new version of openssl with MacOS causing that problem.

### Task logs

/usr/bin/openssl pkcs12 -in /Users/myuser/buildagent/_work/_temp/apple-distribution.p12 -nokeys -passin pass:****** | /usr/bin/openssl x509 -noout -fingerprint -subject -dates -nameopt utf8,sep_semi_plus_space
MAC verified OK
SHA256 Fingerprint=00:00:00:00:00:00:00:00:00:......
subject= UID=xxxxxxxxx; CN=Apple Distribution: myfancyorg (apple org id); OU=apple org id; O=MyCompany; C=US
notBefore=Mar  3 13:44:05 2022 GMT
notAfter=Mar  3 13:44:04 2023 GMT
##[error]Error: Unable to find the certificate SHA1 hash and common name (CN). Verify that the certificate is a valid P12.

## Troubleshooting

Installed cert via Xcode so the build process can use it from there

**Comment by Roman-Shchukin** [comment]
Hi @fntc, thanks for information.
Do you mean this old bug: https://github.com/microsoft/azure-pipelines-tasks/issues/11852?

**Comment by Wackymax** [comment]
Running into the same problem

**Comment by hashtobewild** [comment]
As mentioned before (in the old bug ticket), the openssl output format has changed, i.e.:

```
SHA256 Fingerprint=...
```

in stead of

```
SHA1 Fingerprint=
```

So, as a crayon-level work-around, I added the commands it looked like it was running to my pipeline in a script block. Something like:

```
- script: |
    /usr/bin/security create-keychain -p xxx /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain
    /usr/bin/security set-keychain-settings -lut 21600 /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain/usr/bin/security set-keychain-settings -lut 21600 /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain
    /usr/bin/security unlock-keychain -p xxx /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain
    /usr/bin/security import /Users/someuser/secretfiles/Certificates.p12 -P 'somepassword_from_a_previously_working_run' -A -t cert -f pkcs12 -k /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain
    /usr/bin/security list-keychain -d user
    /usr/bin/security list-keychain -d user -s /Users/someuser/myagent/_work/_temp/ios_signing_temp.keychain /Users/someuser/Library/Keychains/login.keychain-db
    /usr/bin/security list-keychain -d user
  displayName: 'Manual Cert Install'
``` 

I manually copied the certificate file to the `/Users/someuser/secretfiles` directory.

Yes, I know it sucks and there are a fair number of folks rolling their eyes at me right now, but... I sincerely don't care. It got my pipelines working ;)

I hope someone can use it as inspiration to come up with a better solution... or an example of what not to do.

**Comment by namhi** [comment]
Running into the same problem on MacOS 13

**Comment by DaniRvd** [comment]
Same issue here after upgrading to MacOS 13

**Comment by abdulla-allaith** [comment]
Same issue, MacOS 13

**Comment by abdulla-allaith** [comment]
Any update on this issue?
It's breaking all iOS pipelines.

**Comment by Gakk** [comment]
Upgraded to MacOS 13 today (prior to reading this), and the pipeline stopped working 😟

**Comment by Gakk** [comment]
@Roman-Shchukin:

> Do you mean this old bug: #11852?

It seems that the old bug focused on change in output, expecting `/` as separator and getting comma (`,`):

> Expected: subject= /UID=XXXXXXXXX/CN=iPhone Developer: Test Test (XXXXXXXXX) [...]
> Received: subject=UID = XXXXXXXXX, CN = iPhone Developer: Test Test (XXXXXXXXX) [...]

This seems not to be the case now, as both working MacOS 12 and failing MacOS 13 uses semicolon (`;`), so parsing seems to work:

**MacOS 12:**

> MAC verified OK
> SHA1 Fingerprint=XX:XX:XX [...]
> subject= UID=XXXXXXXXXX; CN=iPhone Distribution: Company (XXXXXXXXXX); OU=XXXXXXXXXX; O=Company; C=US

**MacOS 13:**

> MAC verified OK
> SHA256 Fingerprint=XX:XX:XX [...]
> subject= UID=XXXXXXXXXX; CN=iPhone Distribution: Company (XXXXXXXXXX); OU=XXXXXXXXXX; O=Company; C=US

I therefore believe that this is a **new** issue, and **not** related to #11852

**Comment by Roman-Shchukin** [comment]
Hi @abdulla-allaith! We are already working on this issues.

**Comment by ismayilov-ismayil** [comment]
the problem is solved with this PR [https://github.com/microsoft/azure-pipelines-tasks/pull/17110](https://github.com/microsoft/azure-pipelines-tasks/pull/17110) 
and now awaiting deployment

**Comment by mstrent** [comment]
We are also blocked by this issue. Hoping for a quick turnaround on releasing the fix!

**Comment by mstrent** [comment]
@ismayilov-ismayil is there any way to get a rough timeline estimate on when this fix might be rolled out and usable? We need to understand whether we should wait or start something drastic like rebuilding our Mac build server.

**Comment by Gakk** [comment]
I see that the task _"Install Apple Certificate"_ within the last few days was updated from the failing versjon _2.200.0_ to a new version _2.212.0_ that also fails.

It is starting to hurt a lot not being able to push new updates of our iOS apps 😟

But it looks like the PR #17195 targets release branch `m213`, and hopefully that means a new release 2.213.0 very soon 😇 ?

**Comment by Gakk** [comment]
When running the build today version _2.213.0_ is used, and the task now succeeds 🥳

@ismayilov-ismayil, I believe this case now can be closed 🙂

**Comment by jdwirrenga** [comment]
@Gakk is 2.213.0 referring to the version of the build agent? We tried 2.213 and 2.214 and neither made any difference. It seems to me the build agent and the tasks are independent of one another. 

What do all the release branches in this repo correspond to? When the "awaiting deployment" tag is taken off and this fix is "released" how do I look at my task in Azure DevOps and know whether it has the version with the changes in it or not? I can't see any way for us to know what version of this repo's code we are running locally, or how we would consume changes as they are released.

**Comment by ismayilov-ismayil** [comment]
Hi @jdwirrenga 
could you please provide the version of this task by looking through pipeline agent? Or you can provide your organization name on azuredevops, I can check on behalf of you

**Comment by jdwirrenga** [comment]
@ismayilov-ismayil  we are on-prem. So in this instance, how could we get this task's version bumped from 2.170.1 to 2.213.0 or higher?
![image](https://user-images.githubusercontent.com/72875541/207358326-a88aa721-1197-4480-b90e-2c3b82772d6e.png)

**Comment by bpeltonc** [comment]
@alenros Are you able to provide any further instruction on uploading the newly created task to my instance?

**Comment by jdwirrenga** [comment]
@alenros I did as you said and compared this InstallAppleCertificateV2 folder to the one on the server, and they looked mostly the same. We tried simply replacing what was on the server with the files from master branch, restarted the agent and the server, but it still said 2.170.1 for the version when I ran the pipeline. Am I missing an important file or step here?

Now I am going through the steps to publish an extension that is just a copy of this, but that seems like the very roundabout way to do this.

**Comment by bpeltonc** [comment]
@ismayilov-ismayil Our pipeline is also running v2.170.1 for this task and failing with the same error. How can we make it use a newer version with the fix applied?

**Comment by jdwirrenga** [comment]
@bpeltonc I have found a solution. As @alenros suggested, close the repo and build the task with the command above. In the root of this repo, there will be a _build folder. When you are ready to upload the new version, you will run something like ```tfx build tasks upload --task-path "./_build/Tasks/InstallAppleCertificateV2" --service-url "https://youronpremurl"```. All you will need is a PAT and Project Collection Administrator permissions to do this.

Additionally, run ```npm install``` and ```npm install tfx-cli -g``` once you check out the repo. There is a [blog post on automating this that I cheated off of](https://jessehouwing.net/adding-visual-studio-2022-to-azure-devops-server-2020/)

**Comment by bpeltonc** [comment]
@jdwirrenga Thanks for the help! That got it working for me.

**Comment by alenros** [comment]
@jessehouwing has this [neat repo](https://github.com/jessehouwing/azure-pipelines-tasks-zips/releases/tag/m215) with everything already built

**Comment by jessehouwing** [comment]
And the code to make them into an extension for easier installation.

**Comment by jessehouwing** [comment]
Is there a logical set of xcode tasks that would be worth turning into an extension?

**Comment by jessehouwing** [comment]
I just added xcode to the extension builds, if anyone can verify this works for them...

https://github.com/jessehouwing/azure-pipelines-tasks-zips/releases/download/m215/_jessehouwing.Apple-Xcode.vsix

**Comment by jessehouwing** [comment]
This extension should install the latest version of the tasks into your TFS / Azure DevOps Server.

The tasks have `-sxs` appended to their original name (extensions can't overwrite the in-the-box tasks). 

In UI builds and classic release you can recognize them by the `(Side-by-side)` in their name.

https://marketplace.visualstudio.com/items?itemName=jessehouwing.Apple-Xcode

**Comment by mina4gerges** [comment]
any updates ?

**Comment by jessehouwing** [comment]
> any updates ?

If previous situations are any indication, wait for 2022.1 for a fix and for no fix to come to older versions.

**Comment by nzcodarnoc** [comment]
I just updated my self hosted agent on my MacOS M1, and this bug is still present. :(

**Comment by tpgmeligmeyling** [comment]
@nzcodarnoc 
Same, and I expect more people to run into this issue now. I upgraded as I got errors that starting April 2023, all iOS apps submitted to the App Store must be built with the iOS 16.1 SDK or later, included in Xcode 14.1 or later. For Xcode 14, MacOS13 is needed. I suspect more people to upgrade when Xcode 14.1 IS required. 

@jessehouwing 
Thank you, adding https://marketplace.visualstudio.com/items?itemName=jessehouwing.Apple-Xcode to my on-premise DevOps server and using environment variable AZP_AGENT_DOWNGRADE_DISABLED=true was a fast way of getting my pipelines running again!
